xds/src/main/java/io/grpc/xds/internal/sds/SslContextProviderSupplier.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.xds.EnvoyServerProtoData.UpstreamTlsContext;
+import io.netty.handler.ssl.SslContext;
+
+/**
+ * Enables the CDS policy to initialize this object with the received {@link UpstreamTlsContext} &
+ * communicate it to the consumer i.e. {@link SdsProtocolNegotiators.ClientSdsProtocolNegotiator}
+ * to lazily evaluate the {@link SslContextProvider}. The supplier prevents credentials leakage in
+ * cases where the user is not using xDS credentials but the CDS policy contains a non-default
+ * {@link UpstreamTlsContext}.
+ */
+public final class SslContextProviderSupplier implements Closeable {
+
+  private final UpstreamTlsContext upstreamTlsContext;
+  private final TlsContextManager tlsContextManager;
+  private SslContextProvider sslContextProvider;
+  private boolean shutdown;
+
+  public SslContextProviderSupplier(
+      UpstreamTlsContext upstreamTlsContext, TlsContextManager tlsContextManager) {
+    this.upstreamTlsContext = upstreamTlsContext;
+    this.tlsContextManager = tlsContextManager;
+  }
+
+  public UpstreamTlsContext getUpstreamTlsContext() {
+    return upstreamTlsContext;
+  }
+
+  @VisibleForTesting public boolean isShutdown() {
+    return shutdown;
+  }
+
+  /** Updates SslContext via the passed callback. */
+  public synchronized void updateSslContext(final SslContextProvider.Callback callback) {
+    checkNotNull(callback, "callback");
+    if (sslContextProvider == null && !shutdown) {
+      sslContextProvider =
+          tlsContextManager.findOrCreateClientSslContextProvider(upstreamTlsContext);
+    }
+    // we want to increment the ref-count so call findOrCreate again...
```

Comment:
If the ref-count is  incremented more than once, how it will be released by `close()`?

Reply:
If you see the anonymous `Callback` created on line 64, in each method I call `releaseClientSslContextProvider` (lines 69 and 75) which decrements the ref-count (after the `SslContextProvider` calls one of the callback methods) so that leaves the ref-count to 1 from the `findOrCreate` on line 58 which is the expected value when `close` is called to decrement to 0 and release the object.

The Supplier class achieves lazy initialization of `SslContextProvider` on the client side to solve this problem: when a CDS balancer receives a CDS response containing UpstreamTlsContext the code was earlier creating/allocating an SslContextProvider even if the channel didn't use XdsCredentials (i.e. authorizing use of xDS supplied certs). In that case the SslContextProvider was going ahead and sending CallCreds on the wire even if the user had not authorized use of XdsCreds. The problem is solved by using the Supplier which does lazy initialization of the SslContextProvider i.e. on the first actual use of the SslContextProvider it will allocate the object. The ref-count of 2 reflects the original creation (that would have happened in the CDS LB) and the current use.

Reply:
if you consider the pre-Supplier behavior it is something like this:

1) control plane sends a CDS response with a certain UpstreamTlsContext value
2) when the CDS LoadBalancer receives that, it creates an SslContextProvider with that UpstreamTlsContext value and as part of that creation the SslContextProvider does the initial heavy lifting of contacting the CA server, getting the certificates and making the SslContext ready to be consumed by the actual handshaker of a subchannel. The ref-count gets initialized to 1.
3) When a subchannel is created and engages the ProtocolNegotiator it asks for the same SslContextProvider (based on how the Attribute is passed down) and because this SslContextProvider is already initialized the SslContext is just returned by the SslContextProvider. This happens for all the subchannels of a connection and all the connections to any backend in the cluster of that CDS policy. In this interaction the SslContextProvider is accessed (ref-count +1) and immediately released (ref-count -1) after the updateSecret call.
4) At some point the cluster is removed or the UpstreamTlsContext for that cluster changes. The CDS LoadBalancer at that time releases the SslContextProvider and because the ref-count reaches 0 the SslContextProvider is deallocated.

In the above flow because of the initial ref-count of 1 the flow in step 3 does not result into any overhead. With the Supplier modification the initialization becomes lazy so the very first access to the SslContextProvider is delayed but all the subsequent ones in the same CDS are quick.

With what you are suggesting we won't be maintaining the SslContextProvider thru the life of the cluster. But we will be creating and deallocating for every subchannel handshake and that's the overhead we are trying to avoid. For that matter we don't need the Supplier at all if we make `close` a no-op.

Reply:
marking as resolved then